### ISSUE TITLE
Remove need for `api_static` stub files for subpackages and modules

### ISSUE THREAD

**Issue by namurphy** [PlasmaPy/PlasmaPy#1805]
Right now, we often need to [add a stub file](https://docs.plasmapy.org/en/stable/contributing/doc_guide.html#narrative-documentation-guidelines) in `docs/api_static` when adding a new module. This is easy to forget, and I don't think we get an error when it is missing.  I'm wondering if there is a way that we could auto-generate these stub files, or otherwise remove the need for them.

**Comment by rocco8773** [comment]
Our version of `automodsumm` has this configurable with the [`automodapi_generate_module_stub_files`](https://docs.plasmapy.org/en/stable/api_static/plasmapy_sphinx.automodsumm.core.html#confval-automodapi_generate_module_stub_files) configuration variable, however there are complications to using this.  The template for a module stub file is located at [`plasmapy_sphinx/templates/automodsumm/module.rst`](https://github.com/PlasmaPy/PlasmaPy/blob/main/docs/plasmapy_sphinx/templates/automodsumm/module.rst).

Complications:

1. The template stub file is fairly restrictive at the moment, and won't allow for inclusion of notebook galleries.
2. The current module template will generate documentation using `:noindex:` to avoid indexing conflicts.  This has to be done because we don't document all modules using stub files, sometimes a module is documented/indexed in the narrative documentation.  For example, the formulary is only documented/indexed in the narrative documentation.

If you want to incorporate auto generations of stub files, then you have to do one of the following two things:

1. Re-write ALL of the narrative documentation so modules are not indexed and won't conflict with the stub file indexing.  Update `module.rst` with the removal of the `:noindex:` option.
2. Dig into the sphinx rabbit hole that is `automodsumm` and `autosummary` to see if a stub file can be skipped if it's already indexed.  This sounds doable, but it's far more complicated than it appears on the surface.

No. 2 is just not worth our time, and no. 1 will eliminate the possibility for `intersphinx` to link back to custom pages that would include things like jupyter notebooks.  Plus, no. 1 would likely lead to numerous incidents of duplicate documentation for the same module.

---

My humble opinion...as much as I don't like `api_static` either, keeping it is the far simpler solution.  At the moment, it's not worth our limited time to get auto stub file implemented.  There are bigger tofu fish to fry!!!

**Comment by namurphy** [comment]
Agreed...the solutions that we have in mind would take too long to be worthwhile.  I have been wondering if there was an approach that another project is taking that would suit our purposes and not be too much of a pain to implement.  I'd like to keep this issue open because having to add `api_static` files makes it more difficult for newcomers, but we can label it as "on hold" unless we find out about another approach that would be worth the time, or we learn about other Sphinx tooling that we wouldn't have to write ourselves.  Thank you for the detailed response!

**Comment by rocco8773** [comment]
Agreed! However, I'm doubtful that a third party package would be helpful for us because this issue is entangled with the inner workings of both Sphinx's `autosummary` and our own `automodsumm`.

**Comment by StanczakDominik** [comment]
Let it just be known and documented as an official project stance that we pay entirely too much attention to detail in our docs. Are they over engineered? Absolutely, and we're kinda proud of that.

(Which is not to say that I wouldn't want to automate that part away!)

**Comment by namurphy** [comment]
Continuing to brainstorm about this, but I still think we should keep this issue "on hold" for the time being.  The first idea below seems worth considering, while the second idea isn't quite at the point where the benefits and required effort clearly exceed the drawbacks.

**We could possibly add a Nox session to auto-generate a stub file.** The command could be `nox -s make_api_static_stub_file -- src/plasmapy/subpackage/module.py`, which would create the file in `docs/api_static` for `plasmapy.subpackage.module`.  The functionality to do this could be in `tools/`, with a corresponding session in `noxfile.py`.  Probable benefits are that it would be relatively low effort; we'd have more consistency in the stub files; and this would probably save time in the long run.  However, it'd only save time if contributors know about this, and we'd have additional code to maintain.

**We could possibly add a Nox session and CI check to verify that all necessary `api_static` stub files exist.** In doing that we could glob through the `.py` files in `src/plasmapy`, and have a list of modules that do not need an `api_static` page. The upside is that we could have a descriptive error message, but the downside is that we'd have to maintain the list of excluded modules, which could also be confusing.  We'd probably want to exclude top-level subpackages, since those should have narrative docs.  The drawback is that maintaining the list of excluded modules could be confusing and would increase maintenance burden.
